Studio: Add a component that handles a site import

kozer opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/8160

## Proposed Changes

This PR adds import functionality to the studio app.

Keynotes:
1. The implementation has been done on the main process, as it's easier to handle files.
2. Import functionality is structured as follows :
    a. **A backup handler:** A class responsible for handling the different compressed format
    b. **A backup manager:** Responsible for running through the validators, and selecting the appropriate importer
    b. **Validators:** For now, it's only for the Jetpack Backup format, but we can add as many validators as we want. The validator is responsible for determining, based on the files, if each type of format meets its different constraints, and preparing the data for the importer ( based on files we want to filter out, etc )
    c. **Importers:** For now, it's only for the Jetpack Backup format, but we can add as many importers as we want. The importer is responsible for reading a custom format we have implemented and performing the import. It can further act upon files if needed.


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Pre - test

1. Discrepancy script
```bash
#!/bin/bash

if [ "$#" -ne 2 ]; then
	echo "Usage: $0 <path_to_original_backup> <path_to_imported_site>"
	exit 1
fi

ORIGINAL_BACKUP="$1"
IMPORTED_SITE="$2"
LOG_FILE="import_test_$(date +%Y%m%d_%H%M%S).log"

# Function to compare directories
compare_directories() {
	local dir1="$1"
	local dir2="$2"
	local exclude_pattern="$3"

	diff_output=$(diff -r --exclude="$exclude_pattern" "$dir1" "$dir2")

	if [ -z "$diff_output" ]; then
		echo "No differences found in $dir1" | tee -a "$LOG_FILE"
	else
		echo "Differences found in $dir1:" | tee -a "$LOG_FILE"
		echo "$diff_output" | tee -a "$LOG_FILE"
	fi
}

echo "Starting import test at $(date)" | tee "$LOG_FILE"

# Compare wp-content directories
echo "Comparing wp-content directories..." | tee -a "$LOG_FILE"
compare_directories "$ORIGINAL_BACKUP/wp-content" "$IMPORTED_SITE/wp-content" "*.log"


# Check for studio.json
if [ -f "$ORIGINAL_BACKUP/studio.json" ]; then
	echo "Comparing studio.json..." | tee -a "$LOG_FILE"
	diff "$ORIGINAL_BACKUP/studio.json" "$IMPORTED_SITE/studio.json" | tee -a "$LOG_FILE"
	compare_md5 "$ORIGINAL_BACKUP/studio.json" "$IMPORTED_SITE/studio.json"
else
	echo "studio.json not found in original backup" | tee -a "$LOG_FILE"
fi

echo "Comparison completed. Results logged in $LOG_FILE" | tee -a "$LOG_FILE"
```

2. Patch testing UI:
```typescript
diff --git a/src/components/content-tab-import-export.tsx b/src/components/content-tab-import-export.tsx
index 436c8bcc..30acc637 100644
--- a/src/components/content-tab-import-export.tsx
+++ b/src/components/content-tab-import-export.tsx
@@ -1,6 +1,46 @@
+import { useEffect, useState } from 'react';
+import { getIpcApi } from '../lib/get-ipc-api';
+import { BackupArchiveInfo } from '../lib/import-export/import/types';
+
 interface ContentTabImportExportProps {
 	selectedSite: SiteDetails;
 }
-export function ContentTabImportExport( _props: ContentTabImportExportProps ) {
-	return null;
+export function ContentTabImportExport( { selectedSite }: ContentTabImportExportProps ) {
+	const [ file, setFile ] = useState< File | null >( null );
+
+	const handleFileChange = ( e: React.ChangeEvent< HTMLInputElement > ) => {
+		const selectedFile = e.target.files ? e.target.files[ 0 ] : null;
+		if ( selectedFile ) {
+			setFile( selectedFile );
+		}
+	};
+
+	useEffect( () => {
+		if ( file ) {
+			console.log( `File ready for import: ${ file.name }` );
+		}
+	}, [ file ] );
+
+	const handleSubmit = () => {
+		if ( file ) {
+			try {
+				const backupFile: BackupArchiveInfo = {
+					type: file.type,
+					path: file.path,
+				};
+				getIpcApi().importSite( { id: selectedSite.id, backupFile } );
+			} catch ( error ) {
+				console.error( 'Error importing site:', error );
+			}
+		} else {
+			console.warn( 'No file selected for import' );
+		}
+	};
+
+	return (
+		<div>
+			<input type="file" onChange={ handleFileChange } />
+			<button onClick={ handleSubmit }>Submit</button>
+		</div>
+	);
 }

```


#### Test 1: Files imported successfully
1. Apply the above diff
4. Create a new site.
5. Export a jetpack backup from wpcom.
6. Visit `wp-admin` and check your plugins.
7. Try to import the backup
8. Ensure that by visiting wp-admin, under plugins you see something like below
![2024-07-10T21:12:52,329348617+03:00](https://github.com/Automattic/studio/assets/497103/49adb1ee-b345-4fe2-a59b-9fd979ad88c9)

#### Test 2: Check for discrepancies between the files
1. Run the script above using `./compare.sh <path with downloaded zip>/extracted <path of your site>`
2. Ensure that you see something like below:
```Comparing wp-content directories...
Differences found in /home/wild/Downloads/extracted/wp-content:
Only in /home/wild/Studio/my-bold-website/wp-content: database
Only in /home/wild/Studio/my-bold-website/wp-content: db.php
diff -r '--exclude=*.log' /home/wild/Downloads/extracted/wp-content/index.php /home/wild/Studio/my-bold-website/wp-content/index.php
2c2
< //silence is golden
---
> // Silence is golden.
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-32bit-integer-warnings.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-allowed-redirect-hosts.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-check-theme-availability.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-dns-functions.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-permalinks.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-sqlite.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: 0-thumbnails.php
Only in /home/wild/Downloads/extracted/wp-content/mu-plugins: index.php
Only in /home/wild/Studio/my-bold-website/wp-content/mu-plugins: sqlite-database-integration
Only in /home/wild/Studio/my-bold-website/wp-content/plugins: hello.php
Only in /home/wild/Studio/my-bold-website/wp-content/themes: twentytwentyfour
Only in /home/wild/Studio/my-bold-website/wp-content/themes: twentytwentythree
Only in /home/wild/Studio/my-bold-website/wp-content/themes: twentytwentytwo
Only in /home/wild/Studio/my-bold-website/wp-content: uploads
studio.json not found in original backup
Comparison completed. Results logged in import_test_20240710_213658.log
```
3. Ensure that none of the files included in the extracted backup is missing
## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

kozer committed:
Studio: Add a component that handles a site import

kozer committed:
Update: fix lint errors

kozer committed:
Update: Add some comments

kozer committed:
Update: Minor name refactor

kozer committed:
Update: fix importSite function

kozer committed:
Update: Add validator tests

kozer committed:
Update: fix typo

kozer committed:
Update: export importer manager function for test

kozer committed:
Update: backup handler tests

kozer committed:
Update: Add importer tests

kozer committed:
Update: Add import manager tests

kozer committed:
Merge branch 'trunk' into add/backup_importer

kozer committed:
Update: Renamve files to match project file name convention + fix ipc type

kozer committed:
Update: fix lint errors

kozer committed:
Update: Remove DbConfig as we wont need to provide any credentials

kozer committed:
Update: Extract to a random generated directory

kozer committed:
Update: fix extract name to avoid problems with filenames

kozer committed:
Update: Handle updating site phpVersion

kozer committed:
Update: Remove unecessary methods

wojtekn commented on src/lib/import-export/import/importers/jetpack-importer.ts:
```diff
@@ -0,0 +1,57 @@
+import { Importer, ImporterResult, MetaFileData } from './Importer';
+
+export class JetpackImporter implements Importer {
+	constructor( private backup: BackupContents ) {}
```
As the validator fills out the `BackupContents`, which at this point would be the same for different file formats, could we have a generic importer class e.g. `DefaultImporter`, use it for each format, and extend only if really needed?

kozer replied:
I thought about that originally, however, I decided to do it as is ( for now ) to give the flexibility for the importers to do post-process work, in the middle of handling each step, if needed.
There are the following outcomes as I see it:
- The new importers most likely will not do any extra job, other than the one already implemented, so we will end up having repetitive code - which makes your proposal completely valid, and we should create a `DefaultImporter`.
- The new importers will do **almost the same thing** but with some differences here and there ( inside each step ). That will also create repetitive code, even if we have a default parent class.
- The new importers will handle files **completely differently**, which doesn't make that sense to have a subclass, and the interface is just fine.

Given that, do you think it's better to create a `DefaultImporter` and switch back if needed, or it okay as is?

wojtekn replied:
I think that the first option should be the most valid one. Ideally, we will have one handler and will prepare data before the handler starts its job.

kozer replied:
Done!

wojtekn commented on src/lib/import-export/import/handlers/backup-handler.ts:
```diff
@@ -0,0 +1,69 @@
+	extractFiles( file: BackupArchiveInfo, extractionDirectory: string ): Promise< void >;
+}
+
+export class BackupHandler implements IBackupHandler {
```
When set up that way, shouldn't we do the following?
- Have `BackupHandlerZip` and `BackupHandlerTarGz` implementing the interface
- Have `BackupHandlerFactory` that runs inside `importBackup()` and instantiates correct handles based on the provided file type?

kozer replied:
I thought about it. However, for now, in total we could have 3 potential formats if I'm not missing something:
- gz
- zip
- .wpress

Given that, having a factory may be confusing for anyone looking at the code, ( non experience developers ) especially if we don't expect other formats shortly.
On the other hand, zip and gz are file formats that can be used by different importers, and thus I took the time to create those interfaces etc.
What do you think?

wojtekn replied:
It still makes sense to follow a single responsibility principle and have a class for each, so we could avoid having if/else in each method.

wojtekn commented on src/lib/import-export/import/importers/jetpack-importer.ts:
```diff
@@ -0,0 +1,57 @@
+		}
+	}
+
+	private async handleMetaFile(): Promise< MetaFileData | undefined > {
```
Would it make more sense to name it `read` or `parse` instead of `handle`?

kozer replied:
Yeah, I like `parse`!

kozer commented on src/lib/import-export/import/handlers/backup-handler.ts:
```diff
@@ -0,0 +1,69 @@
+		} else if ( backup.type === 'application/zip' && backupFileExtension === '.zip' ) {
+			return this.listZipContents( backup.path );
+		} else {
+			throw new Error( 'Unsupported file format. Only .gz and .zip files are supported.' );
```
Note: We can add more supported formats in the future.

kozer commented on src/lib/import-export/import/importers/index.ts:
```diff
@@ -0,0 +1,10 @@
+export * from './jetpack-importer';
+
+export const allImporters: { [ key: string ]: new ( contents: BackupContents ) => Importer } = {
+	JetpackValidator: JetpackImporter,
```
Note: More importers can be added in the future.

kozer commented on src/lib/import-export/import/types.ts:
```diff
@@ -0,0 +1,17 @@
+		[ key: string ]: string[]; // 'uploads', 'plugins', 'themes'
+	};
+	metaFile?: string;
+}
```
Note: Format that all importers will be using.

wojtekn commented on src/lib/tests/import-export/import/handlers/backup-handler.test.ts:
```diff
@@ -0,0 +1,124 @@
+import zlib from 'zlib';
+import AdmZip from 'adm-zip';
+import * as tar from 'tar';
+import { BackupHandler } from '../../../../import-export/import/handlers/backup-handler';
```
Could we place tests closer to the component we are testing, similarly to other files?

kozer replied:
I'll move them inside `import` folder. I wanted to have all tests in one place ( related to `lib` ), but I get your point

wojtekn reviewed (commented):
It looks like a great start @kozer ! I left some comments related to architecture.

kozer commented on src/ipc-handlers.ts:
```diff
@@ -99,6 +103,28 @@ export async function getInstalledApps( _event: IpcMainInvokeEvent ): Promise< I
+) {
+	const site = SiteServer.get( id );
+	if ( ! site ) {
+		throw new Error( 'Site not found.' );
```
Note: New site creation will be handled from the UI.

wojtekn replied:
Not really from UI, but Studio would create an empty site just before it executes the import logic, correct?

kozer replied:
Yeah, by "from UI" I meant the renderer process ( the React part ).

wojtekn commented:
@kozer I know it's early stage, but I tried importing Jetpack Backup and got this error in the console:

```
Backup import failed: Unsupported file format. Only .gz and .zip files are supported.
Error importing site Unsupported file format. Only .gz and .zip files are supported.
```

Given the current implementation, what would be the plan to support importing the MySQL dump file alone, in either unpacked form as `.sql` or packed as `.zip` or `.gz`?

kozer commented:
> @kozer I know it's early stage, but I tried importing Jetpack Backup and got this error in the console:
> 
> ```
> Backup import failed: Unsupported file format. Only .gz and .zip files are supported.
> Error importing site Unsupported file format. Only .gz and .zip files are supported.
> ```
> 
> Given the current implementation, what would be the plan to support importing the MySQL dump file alone, in either unpacked form as `.sql` or packed as `.zip` or `.gz`?

I exported a site from an atomic WPCOM site I have ( from Jetpack menu )
Can you share more instructions on how you ended up having an error? I need to know the following:
1. Where did you get your backup?
2. What is your backup extension? ( It may be better to avoid having a check for `gz` or `.zip`  extension and rely only on type. After all, the extension means nothing )

I'm not handling `.sql` as is, but it's an easy fix. I'll do that now.

kozer committed:
Update: Move tests closer to implementation

wojtekn commented:
> I exported a site from an atomic WPCOM site I have ( from Jetpack menu )
> Can you share more instructions on how you ended up having an error? I need to know the following:
> 
> Where did you get your backup?
> What is your backup extension? ( It may be better to avoid having a check for gz or .zip extension and rely only on type. After all, the extension means nothing )

I've got it from the Activity Log. As I thought it may be system-specific error, I added some debug logic in the handler:

```
		console.error( backup.type );
		console.error( backupFileExtension );
```

The output looks as follows:

```
application/x-gzip
.gz
```

> I'm not handling .sql as is, but it's an easy fix. I'll do that now.

I was more interested in the plan for handling it. I didn't intend to request immediate implementation.

kozer commented:
It seems it's `application/x-gzip`. For me its just `application/gzip` ( I also exported from activity log ). 
Looking now.

UPDATE:
Seems that `x-gzip` it's there for compatibility reasons ( [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) )

kozer committed:
Update: refactor based on comments

- Replace `rmdir` with `rm` as we got a warning ( due to recursive: true )
- Add slqValidator and importer
- Create a DefaultImporter and let all others extend that
- Create a backup handler factory to hide handler logic per type ( zip,
  sql, tar.gz )
- Fix and update tests

kozer commented:
Update PR based on latest discussions

kozer committed:
Update: test comments on how to run them

kozer committed:
Merge branch 'trunk' into add/backup_importer

wojtekn commented on src/lib/import-export/import/importers/sql-importer.ts:
```diff
@@ -0,0 +1,8 @@
+import { BackupContents } from '../types';
+import { DefaultImporter } from './Importer';
+
+export class SqlImporter extends DefaultImporter {
```
As those importers don't differ, and we don't know if they will, should we just use `DefaultImporter` in all cases and avoid extending the base one?

kozer replied:
Done! :smile:

wojtekn commented on src/lib/import-export/import/importers/Importer.ts:
```diff
@@ -0,0 +1,67 @@
+	}
+
+	protected async importDatabase(): Promise< void > {
+		// empty in DefaultImporter
```
Each importer should import SQL files if they are present. As we will separately implement that part, should we adjust this comment to make it clearer that we will implement it under another issue?

kozer replied:
Done!

wojtekn commented on src/lib/import-export/import/handlers/backup-handler-factory.ts:
```diff
@@ -0,0 +1,72 @@
+import { BackupHandlerTarGz } from './backup-handler-tar-gz';
+import { BackupHandlerZip } from './backup-handler-zip';
+
+export interface IBackupHandler {
```
As the `BackupHandler` name is free now, can we use this one for the interface?

wojtekn commented:
Feature-wise, it worked for me now. I was able to import a site from Jetpack Backup downloaded from the Activity Log of my self-hosted site.

kozer committed:
Update: Minor changes based on comments ( renames etc ).

kozer commented:
Updated PR based on latest comments!

fluiddot commented on src/lib/import-export/import/import-manager.ts:
```diff
@@ -0,0 +1,51 @@
+	validators: Validator[],
+	importers: { [ key: string ]: new ( backup: BackupContents ) => Importer }
+): Promise< ImporterResult > {
+	const extractionDirectory = await fsPromises.mkdtemp( path.join( os.tmpdir(), 'studio_backup' ) );
```
I'd suggest including a UUID in the temporary path to allow multiple import processes simultaneously.

kozer replied:
You mean an extra id?
Because from the docs of `mkdtemp`:
```
The fs.mkdtemp() method is used to create a unique temporary directory. The folder name is 
created by appending 6 randomly generated characters behind a prefix string.
It is also possible to create the temporary directory inside a folder by using a separator after the folder path. ```

fluiddot replied:
Ah, true. I overlooked that function `mkdtemp` already appends a random id. I confused this with another implementation I did in E2E tests where I appended a UUID for a temp folder ([reference](https://github.com/Automattic/studio/blob/ca29e98b1aa7f28210b93cc2b5c9c6c11e96bc2e/e2e/e2e-helpers.ts#L18)). Forget my comment 😅 .

fluiddot commented on src/lib/import-export/import/handlers/backup-handler-factory.ts:
```diff
@@ -0,0 +1,72 @@
+	private static isZip( file: BackupArchiveInfo ): boolean {
+		return (
+			this.zipTypes.includes( file.type ) &&
+			this.zipExtensions.some( ( ext ) => file.path.endsWith( ext ) )
```
I wonder if it's redundant to check the file extension if we are already checking the MIME type of the file. WDYT?

kozer replied:
It is a bit,but originally I thought that many users are not familiar with MIME types, but rather with file extensions. Checking ext as well will make more clear for those users what we want.
What do you think?

fluiddot replied:
Definitely, we should filter by extension in the file dialog when importing a file, but I presume that it will be handled in a different place. If I understand this part correctly, these conditions are mainly used to determine the handler to import the file, which might be agnostic to extension. In any case, I don't have a strong preference to remove the extension condition so we can keep it. My comment is mainly driven by the case of dropping a file. Probably this is an edge case, but if the user drops a backup file without extension, should we process it if it has a valid MIME type or discard it?

kozer replied:
>  Probably this is an edge case, but if the user drops a backup file without extension, should we process it if it has a valid MIME type or discard it?

Another problem is the following:
- Normal gz files, ( not tar ), have the same mime type as tar.gz. Currently, as is, we won't try to handle those files ( which would fail the process, down the line ).

Given that, let's keep the extension for now, as it's not easy to determine solely on the MIME type how to handle those files

fluiddot replied:
Ah, good point. In that case, let's keep the condition that checks the file extension 👍 .

fluiddot commented on src/ipc-handlers.ts:
```diff
@@ -99,6 +103,28 @@ export async function getInstalledApps( _event: IpcMainInvokeEvent ): Promise< I
+			} );
+		}
+	} catch ( e ) {
+		console.log( 'Error importing site', ( e as Error ).message );
```
When we use this IPC handler in the UI we might need to return the error to display an error message.

kozer replied:
This function is subject to change a little bit, as fit, when the UI is going to be implemented.
However, this is something that is going to be needed either way, so I 'll update, and return the error.

fluiddot reviewed (commented):
I haven't fully reviewed and tested the PR but sharing preliminary comments.

kozer committed:
Update: Rethrow error for UI to be able to act on it

fluiddot commented on src/lib/import-export/import/import-manager.ts:
```diff
@@ -0,0 +1,51 @@
+	allFiles: string[],
+	extractionDirectory: string,
+	validators: Validator[],
+	importers: { [ key: string ]: new ( backup: BackupContents ) => Importer }
```
Nitpick: It would be great to have this defined as a type in `lib/import-export/import/types.ts`. WDYT?

kozer replied:
Updated! Thanks for the comment!

fluiddot commented on src/lib/import-export/import/validators/Validator.ts:
Usually, JS file names are downcase, what do you think about changing this to `validator.ts`?

fluiddot commented on src/lib/import-export/import/importers/Importer.ts:
Usually, JS file names are downcase, what do you think about changing this to `importer.ts`?

fluiddot commented on src/lib/import-export/import/import-manager.ts:
```diff
@@ -0,0 +1,51 @@
+): Importer | null {
+	for ( const validator of validators ) {
+		if ( validator.canHandle( allFiles ) ) {
+			const importerClass = importers[ validator.constructor.name ];
```
If each validator runs a specific importer, I wonder if we could make a different association instead of using validator's constructor name. Here's an example where the association is made when invoking `importBackup` function:

```patch
diff --git forkSrcPrefix/src/lib/import-export/import/import-manager.ts forkDstPrefix/src/lib/import-export/import/import-manager.ts
index a9dcbd0c0d4c3d946476b0692d07c6a2f5b747ae..532a5b56b06484dac0cf04373839b4ad71b94546 100644
--- forkSrcPrefix/src/lib/import-export/import/import-manager.ts
+++ forkDstPrefix/src/lib/import-export/import/import-manager.ts
@@ -6,19 +6,20 @@ import { Importer, ImporterResult } from './importers/Importer';
 import { BackupArchiveInfo, BackupContents } from './types';
 import { Validator } from './validators/Validator';
 
+export interface ImporterOption {
+	validator: Validator;
+	importer: new ( backup: BackupContents ) => Importer;
+}
+
 export function selectImporter(
 	allFiles: string[],
 	extractionDirectory: string,
-	validators: Validator[],
-	importers: { [ key: string ]: new ( backup: BackupContents ) => Importer }
+	options: ImporterOption[]
 ): Importer | null {
-	for ( const validator of validators ) {
+	for ( const { validator, importer } of options ) {
 		if ( validator.canHandle( allFiles ) ) {
-			const importerClass = importers[ validator.constructor.name ];
-			if ( importerClass ) {
-				const files = validator.parseBackupContents( allFiles, extractionDirectory );
-				return new importerClass( files );
-			}
+			const files = validator.parseBackupContents( allFiles, extractionDirectory );
+			return new importer( files );
 		}
 	}
 	return null;
@@ -27,14 +28,13 @@ export function selectImporter(
 export async function importBackup(
 	backupFile: BackupArchiveInfo,
 	sitePath: string,
-	validators: Validator[],
-	importers: { [ key: string ]: new ( backup: BackupContents ) => Importer }
+	options: ImporterOption[]
 ): Promise< ImporterResult > {
 	const extractionDirectory = await fsPromises.mkdtemp( path.join( os.tmpdir(), 'studio_backup' ) );
 	try {
 		const backupHandler = BackupHandlerFactory.create( backupFile );
 		const fileList = await backupHandler.listFiles( backupFile );
-		const importer = selectImporter( fileList, extractionDirectory, validators, importers );
+		const importer = selectImporter( fileList, extractionDirectory, options );
 
 		if ( importer ) {
 			await backupHandler.extractFiles( backupFile, extractionDirectory );
diff --git forkSrcPrefix/src/ipc-handlers.ts forkDstPrefix/src/ipc-handlers.ts
index d7014d5223965139d627ffa483f40a8615a7c298..702b07c9c62a17bfbd9d15067b2bb26e6fb856db 100644
--- forkSrcPrefix/src/ipc-handlers.ts
+++ forkDstPrefix/src/ipc-handlers.ts
@@ -19,10 +19,10 @@ import { downloadSqliteIntegrationPlugin } from '../vendor/wp-now/src/download';
 import { LIMIT_ARCHIVE_SIZE } from './constants';
 import { isEmptyDir, pathExists, isWordPressDirectory, sanitizeFolderName } from './lib/fs-utils';
 import { getImageData } from './lib/get-image-data';
-import { importBackup } from './lib/import-export/import/import-manager';
-import { allImporters } from './lib/import-export/import/importers';
+import { ImporterOption, importBackup } from './lib/import-export/import/import-manager';
+import { DefaultImporter } from './lib/import-export/import/importers';
 import { BackupArchiveInfo } from './lib/import-export/import/types';
-import { allValidators } from './lib/import-export/import/validators';
+import { JetpackValidator, SqlValidator } from './lib/import-export/import/validators';
 import { isErrnoException } from './lib/is-errno-exception';
 import { isInstalled } from './lib/is-installed';
 import { getLocaleData, getSupportedLocale } from './lib/locale';
@@ -113,7 +113,11 @@ export async function importSite(
 	}
 	const sitePath = site.details.path;
 	try {
-		const result = await importBackup( backupFile, sitePath, allValidators, allImporters );
+		const defaultOptions: ImporterOption[] = [
+			{ validator: new JetpackValidator(), importer: DefaultImporter },
+			{ validator: new SqlValidator(), importer: DefaultImporter },
+		];
+		const result = await importBackup( backupFile, sitePath, defaultOptions );
 		if ( result?.meta?.phpVersion ) {
 			await updateSite( event, {
 				...site.details,
@@ -122,7 +126,7 @@ export async function importSite(
 		}
 	} catch ( e ) {
 		Sentry.captureException( e );
-		throw e
+		throw e;
 	}
 }
 
diff --git forkSrcPrefix/src/lib/import-export/import/validators/index.ts forkDstPrefix/src/lib/import-export/import/validators/index.ts
index a34fb9a9964a3a5da8d15794b486c83976521d15..14ee5e75bcb6e2a809984ce2aea763e232fb134a 100644
--- forkSrcPrefix/src/lib/import-export/import/validators/index.ts
+++ forkDstPrefix/src/lib/import-export/import/validators/index.ts
@@ -1,8 +1,3 @@
-import { Validator } from './Validator';
-import { JetpackValidator } from './jetpack-validator';
-import { SqlValidator } from './sql-validator';
-
 export * from './Validator';
 export * from './jetpack-validator';
-
-export const allValidators: Validator[] = [ new SqlValidator(), new JetpackValidator() ];
+export * from './sql-validator';

```

kozer replied:
I like the change! On it!

kozer replied:
Updated!

kozer committed:
Update: fix lint errors

fluiddot commented on src/lib/import-export/import/validators/sql-validator.ts:
```diff
@@ -0,0 +1,30 @@
+
+export class SqlValidator implements Validator {
+	canHandle( fileList: string[] ): boolean {
+		return fileList.length === 1 && fileList[ 0 ].endsWith( '.sql' );
```
Based on the PR's testing instructions, I'm not sure if this part is ready and should be tested. In any case, I'm out of curiosity why we only allow importing a single SQL file. I presume it's related to the expectation of having a single big file with all SQL statements, but I wonder if we should also support a zipped file that contains multiple SQL files. WDYT?

This comment is tangentially related to 8282-gh-Automattic/dotcom-forge, as it mentions support `.zip` and `.tar.gz` for single `.sql` files.

kozer replied:
> but I wonder if we should also support a zipped file that contains multiple SQL files. WDYT?

As this is already a big one, after discussion with Wojtek, I opened a new ticket: https://github.com/Automattic/dotcom-forge/issues/8282

fluiddot replied:
Sound good, let's move the discussion to that ticket.

fluiddot commented on src/lib/import-export/import/validators/sql-validator.ts:
```diff
@@ -0,0 +1,30 @@
+		const extractedBackup: BackupContents = {
+			extractionDirectory: extractionDirectory,
+			sqlFiles: [],
+			wpContent: {
+				uploads: [],
+				plugins: [],
+				themes: [],
+			},
```
Do we need to define `wp-content` folders if this validator is only meant to be used for SQL files?

kozer replied:
Thanks for your comment! I did that, to keep the API the same between the validators and to keep a common interface, that we give to the importer.

fluiddot replied:
Ok, makes sense. I understand we'll keep `wpContent` as part of the interface, but could we remove the subfolders?
```suggestion
			wpContent: {},
```

kozer replied:
I'm not fan of improperly handled types tbh, as one doesn't have the whole picture by just looking at it.
Instead, I updated the wpContent object to be enforced. 
Thanks for your comment!

fluiddot commented on src/lib/import-export/import/handlers/backup-handler-factory.ts:
```diff
@@ -0,0 +1,71 @@
+		);
+	}
+
+	private static isSql( file: BackupArchiveInfo ): boolean {
+		return (
+			this.sqlTypes.includes( file.type ) &&
+			this.sqlExtensions.some( ( ext ) => file.path.endsWith( ext ) )
+		);
+	}
```
I tried to import one of the `.sql` files from a Jetpack backup export, and these conditions weren't met. I debugged the `type` and it's empty. This makes me think that in case we can't determine the MIME type, we should only heck the extension. WDYT?

fluiddot replied:
Potential change:
```suggestion
	private static isSql( file: BackupArchiveInfo ): boolean {
		return (
			( this.sqlTypes.includes( file.type ) || ! file.type ) &&
			this.sqlExtensions.some( ( ext ) => file.path.endsWith( ext ) )
		);
	}
```

kozer replied:
This is interesting, as this doesn't happening on linux. I see the correct SQL mime type. Let's submit your changes, and test it more in depth, when we have the SQL import part ready.

fluiddot replied:
So far, I've tested multiple SQL files generated from different sources and all set the `type` as empty. I wonder if you could share the SQL file you used, just in case it's related to the file format. Thanks!

kozer committed:
Update src/lib/import-export/import/handlers/backup-handler-factory.ts

Co-authored-by: Carlos Garcia <fluiddot@gmail.com>

kozer committed:
Update: minor updates based on comments

fluiddot commented on src/lib/import-export/tests/import/handlers/backup-handler.test.ts:
```diff
@@ -0,0 +1,216 @@
+	} );
+
+	describe( 'listFiles', () => {
+		it( 'should extract files from a gzip archive', async () => {
```
I noticed that some of the tests under this section are related to extracting files, not the `listFiles` function. Only `should list a single SQL file` is calling the `listFiles` function.

fluiddot replied:
I can help with this as I'm working on similar unit tests for 8282-gh-Automattic/dotcom-forge.

fluiddot replied:
Here's a potential path to apply:

```patch
diff --git forkSrcPrefix/src/lib/import-export/tests/import/handlers/backup-handler.test.ts forkDstPrefix/src/lib/import-export/tests/import/handlers/backup-handler.test.ts
index 2dd04d8a360c070a651bcfc7c547bfe5d611547f..eabf9b6cfc5c597ea0b77f3bb30299fbb832cb9b 100644
--- forkSrcPrefix/src/lib/import-export/tests/import/handlers/backup-handler.test.ts
+++ forkDstPrefix/src/lib/import-export/tests/import/handlers/backup-handler.test.ts
@@ -1,7 +1,6 @@
 // To run tests, execute `npm run test -- src/lib/import-export/tests/import/handlers/backup-handler-factory.test.ts`
 import fs from 'fs';
 import path from 'path';
-import zlib from 'zlib';
 import AdmZip from 'adm-zip';
 import * as tar from 'tar';
 import { BackupHandlerFactory } from '../../../import/handlers/backup-handler-factory';
@@ -66,71 +65,41 @@ describe( 'BackupHandlerFactory', () => {
 	} );
 
 	describe( 'listFiles', () => {
-		it( 'should extract files from a gzip archive', async () => {
+		const archiveFiles = [
+			'index.php',
+			'wp-content/plugins/hello.php',
+			'wp-content/themes/twentytwentyfour/theme.json',
+			'wp-content/uploads/2024/07/image.png',
+		];
+
+		it( 'should list files from a gzip archive', async () => {
 			const archiveInfo: BackupArchiveInfo = {
 				path: '/path/to/backup.tar.gz',
 				type: 'application/gzip',
 			};
 			const handler = BackupHandlerFactory.create( archiveInfo );
-			const extractionDirectory = '/tmp/extracted';
 
-			( path.extname as jest.Mock ).mockReturnValue( '.gz' );
-			( fs.createReadStream as jest.Mock ).mockReturnValue( {
-				pipe: jest.fn().mockReturnThis(),
-				on: jest.fn().mockImplementation( ( event, callback ) => {
-					if ( event === 'finish' ) {
-						callback();
-					}
-					return this;
-				} ),
-			} );
-			( zlib.createGunzip as jest.Mock ).mockReturnValue( {
-				pipe: jest.fn().mockReturnThis(),
-			} );
-			( tar.extract as unknown as jest.Mock ).mockReturnValue( {
-				on: jest.fn().mockImplementation( ( event, callback ) => {
-					if ( event === 'finish' ) {
-						callback();
-					}
-					return this;
-				} ),
+			jest.spyOn( tar, 't' ).mockImplementation( ( { onReadEntry } ) => {
+				archiveFiles.forEach( ( path ) => onReadEntry?.( { path } as tar.ReadEntry ) );
 			} );
 
-			await expect(
-				handler.extractFiles( archiveInfo, extractionDirectory )
-			).resolves.not.toThrow();
+			await expect( handler.listFiles( archiveInfo ) ).resolves.toEqual( archiveFiles );
 		} );
 
-		it( 'should extract files from a zip archive', async () => {
+		it( 'should list files from a zip archive', async () => {
 			const archiveInfo: BackupArchiveInfo = {
 				path: '/path/to/backup.zip',
 				type: 'application/zip',
 			};
 			const handler = BackupHandlerFactory.create( archiveInfo );
-			const extractionDirectory = '/tmp/extracted';
-
-			( path.extname as jest.Mock ).mockReturnValue( '.zip' );
 
-			const mockExtractAllToAsync = jest
-				.fn()
-				.mockImplementation( ( _path, _overwrite, _keepOriginalPermission, callback ) => {
-					callback();
-				} );
-
-			( AdmZip as jest.Mock ).mockImplementation( () => ( {
-				extractAllToAsync: mockExtractAllToAsync,
-			} ) );
-
-			await expect(
-				handler.extractFiles( archiveInfo, extractionDirectory )
-			).resolves.not.toThrow();
+			( AdmZip as jest.Mock ).mockReturnValue( {
+				getEntries: jest
+					.fn()
+					.mockReturnValue( archiveFiles.map( ( file ) => ( { entryName: file } ) ) ),
+			} );
 
-			expect( mockExtractAllToAsync ).toHaveBeenCalledWith(
-				extractionDirectory,
-				true,
-				undefined,
-				expect.any( Function )
-			);
+			await expect( handler.listFiles( archiveInfo ) ).resolves.toEqual( archiveFiles );
 		} );
 
 		it( 'should list a single SQL file', async () => {
@@ -139,7 +108,6 @@ describe( 'BackupHandlerFactory', () => {
 				type: 'application/sql',
 			};
 			const handler = BackupHandlerFactory.create( archiveInfo );
-			( path.extname as jest.Mock ).mockReturnValue( '.sql' );
 			( path.basename as jest.Mock ).mockReturnValue( 'backup.sql' );
 			const result = await handler.listFiles( archiveInfo );
 			expect( result ).toEqual( [ 'backup.sql' ] );
@@ -154,9 +122,16 @@ describe( 'BackupHandlerFactory', () => {
 			};
 			const handler = BackupHandlerFactory.create( archiveInfo );
 			const extractionDirectory = '/tmp/extracted';
-			( path.extname as jest.Mock ).mockReturnValue( '.gz' );
 
-			( tar.x as unknown as jest.Mock ).mockResolvedValue( undefined );
+			( fs.createReadStream as jest.Mock ).mockReturnValue( {
+				pipe: jest.fn().mockReturnThis(),
+				on: jest.fn().mockImplementation( ( event, callback ) => {
+					if ( event === 'finish' ) {
+						callback();
+					}
+					return this;
+				} ),
+			} );
 
 			await expect(
 				handler.extractFiles( archiveInfo, extractionDirectory )
@@ -173,7 +148,6 @@ describe( 'BackupHandlerFactory', () => {
 			};
 			const handler = BackupHandlerFactory.create( archiveInfo );
 			const extractionDirectory = '/tmp/extracted';
-			( path.extname as jest.Mock ).mockReturnValue( '.zip' );
 			const mockExtractAllToAsync = jest
 				.fn()
 				.mockImplementation( ( path, overwrite, keepOriginalPermission, callback ) => {
@@ -200,7 +174,6 @@ describe( 'BackupHandlerFactory', () => {
 			};
 			const handler = BackupHandlerFactory.create( archiveInfo );
 			const extractionDirectory = '/tmp/extracted';
-			( path.extname as jest.Mock ).mockReturnValue( '.sql' );
 			( path.basename as jest.Mock ).mockReturnValue( 'backup.sql' );
 			( fs.promises.copyFile as jest.Mock ).mockResolvedValue( undefined );
 

```

kozer replied:
Updated the tests as requested! Thanks for the effort!

kozer committed:
Update: fix tests

kozer committed:
Update: extract new importer type to types file

kozer committed:
Update: fix lint errors

kozer commented:
@wojtekn , @fluiddot , please when you have a time, let's get a final review, so I can push it today, if possible.
Thanks!

kozer committed:
Merge branch 'trunk' into add/backup_importer

fluiddot reviewed (commented):
LGTM 🎊 ! Awesome work @kozer 🏅 !

sejas reviewed (commented):
@kozer , Great work!

We can consider removing `unzipper` dependency in favor of `adm-zip` or vice versa to avoid having two different dependencies doing the same.
